Elm-analyser integration

hakonrossebo opened the pull request:
This is just a temporary PR to get some initial feedback. Working with @stil4m on https://github.com/stil4m/elm-analyse/issues/77.

This implementation uses elm-analyse and adds issues to the diagnostics collection. To be able to test this:
1. npm install -g elm-analyse
2. run command "Elm-analyse"
3. update/save file to enable linter/diagnostics with updated issues from elm-analyse

Is there any way we can update diagnostics/linter without saving a file? Also, it would be nice to get feedback on when to clean up elm-analyse issues from diagnostics. 

If/when using the elm-analyse web socket/api this could be simplified to fetch info from elm-analyse when saving a file.

hakonrossebo commented:
Yes, i think we need to use the elm-analyse server approach with fast incremental updates. If not we need to only run as a manual elm-analyse command and refresh diagnostics when finished.

If we use elm-analyse server, I think we need to have this option disabled by default.

hakonrossebo commented:
(Note - this is still only for testing/feedback)
I've updated my PR to use websockets, as elm-analyse uses websockets for the live server.

When trying to implement the websockets client, I ended up with the npm ws package. To get typings for ws, I also ran typings install ws. This added some files to typings + generated a typings.json. Is there a better way of doing this with the Typescript version currently in use? Should we use websockets at all, do we prefer any other communications?

Currently, I only have a command to start the elm-analyse process+ws client. I also plan to enable this on startup with a setting.

(Note - forgot to remove "@types/websocket": "0.0.33" from devDependencies)

Krzysztof-Cieslak commented:
`ws` package is fine, I'm pretty sure I've used the same when I've experimented with websockets in F# plugin.

Krzysztof-Cieslak commented:
For the typings - I believe that before we've manually put stuff in the `typings` dir

hakonrossebo commented:
There are still some issues to resolve related to stabililty and functionality, but I think it's ready enough to be released as "experimental support". I've updated the Readme with docs/info here https://github.com/hakonrossebo/vscode-elm/blob/95fd646204fdf523aa3c37a8132771991d4c2de0/README.md

@Krzysztof-Cieslak - There are some commits fixing/improving issues in utils.ts - if you prefer, I can probably take these out and put into a separate PR.

@stil4m - Do you think elm-analyse / this integration is ready? I don't want to publish this before elm-analyse is ready, or if you want to release your atom plugin first.

stil4m commented:
Great work 👍 !

I have a few things Im not sure about, but first I will read through the code (tonight, this weekend). I will make a list of things that we need to consider, and decide on this later (maybe make issues/a project) for it. I'll come back on this.

And again. Thanks for doing this. It is great to see that people need/want this and actually make it. :)

stil4m commented on README.md:
````diff
@@ -149,6 +150,45 @@ You can also configure `elm-format` to run on save by enabling the `elm.formatOn
 }
 ```
+### Elm-analyse integration (Experimental)
+[Elm-analylse](https://github.com/stil4m/elm-analyse) is a tool that allows you to analyse your Elm code, identify deficiencies and apply best practices.
````
Type in the name

hakonrossebo replied:
Fixed in next update.

stil4m commented on README.md:
```diff
@@ -149,6 +150,45 @@ You can also configure `elm-format` to run on save by enabling the `elm.formatOn
+NOTE: *This is currently experimental*
+
+* Elm-analyse is still under development. API will change.
+* There is an issue on Windows where elm-analyse duplicates issues on save
```
This was resolved or not?

hakonrossebo replied:
Removed in next update.

stil4m commented on package.json:
```diff
@@ -197,6 +217,14 @@
+      },
+      {
+        "command": "elm.analyseStop",
+        "title": "Elm: Stop elm-analyse."
```
Maybe remove the period?

hakonrossebo replied:
Fixed in next update.

stil4m commented on src/elmAnalyse.ts:
```diff
@@ -0,0 +1,240 @@
+  private analyseSocket: WebSocket;
+  private analyse: ExecutingCmd;
+  private oc: vscode.OutputChannel = vscode.window.createOutputChannel('Elm Analyse');
+  private messageDescriptionsMap : Map<string, string>
```
I'm not that familiar with Type Script, but shouldn't this line be closed with a semicolon?

hakonrossebo replied:
Fixed in next update.

stil4m commented on src/elmAnalyse.ts:
```diff
@@ -0,0 +1,240 @@
+                let messageInfoFileRange = [0, 0, 0, 0];
+                if (messageTypeInfo.hasOwnProperty("range")) {
+                  messageInfoFileRange = messageTypeInfo.range.real;
+                }
```
It maybe better to extract this logic into a function. That would be the right place to implement the cases where the range is not set on the "range" property of the message. Working on the API docs at the moment, this will become more clear soon.

hakonrossebo replied:
Yes, better to put into function. Will do it in next update.

stil4m commented on src/elmAnalyse.ts:
```diff
@@ -0,0 +1,240 @@
+      });
+      this.analyseSocket.on("error", (e) => {
+        vscode.window.showErrorMessage('Running websocket against Elm-analyse failed. Check if elm-analyse has been configured correctly.');
+        console.log("ERROR", e);
```
You seem not to be consistent using single and double quotes. Is this intentionally? A TS thing?

hakonrossebo replied:
Have not used TS very much myself - I've tried to be consistent, but it's slipping - hope we can have a linter/prettier implementation to help with this later.

stil4m commented on src/elmAnalyse.ts:
```diff
@@ -0,0 +1,240 @@
+          analyseCommand,
+          {
+            fileName: fileName,
+            cmdArguments: ['-s', '-p ' + analysePort],
```
Couldnt you do this: `['-s', '-p', analysePort]`?

hakonrossebo replied:
Thanks - Fixed in next update.

stil4m commented on src/elmAnalyse.ts:
```diff
@@ -0,0 +1,240 @@
+              if (data) {
+                let info = data.toString();
+                this.oc.append(info);
+                if (info.match(/^Found \d* message/) && (this.analyseSocket === undefined || this.analyseSocket.CLOSED == 1)) {
```
You do not have to wait for this message to start the socket. This is started directly on server start.

hakonrossebo replied:
Experienced a few times that the socket was not ready. Hope to improve this by separating logic.

hakonrossebo replied:
Not very experienced with websockets. Do you know if there is a good way to check/know if the server (and the correct server) is running?

stil4m replied:
Not from the top of my head. To check if the correct server is running, we may do a HTTP request prior to connecting returning information of elm-analyse (version number, status, ect). If this call succeeds, then we may know if the server is correctly spawned. But such a mechanism is not in place at the moment.

For now you might better look for the output that matches this: `Listening on http://localhost:3000`.

hakonrossebo replied:
Ok, will try to check the "listening message". Think the status/version information resource will be great in a later version :)

stil4m commented on src/elmAnalyse.ts:
```diff
@@ -0,0 +1,240 @@
+    try {
+      if (editor.document.languageId !== 'elm') {
+        return;
+      }
```
Maybe extract this statement out of the `try` or will this throw an error?

hakonrossebo replied:
Done. Fixed in next update.

stil4m commented on src/elmAnalyse.ts:
```diff
@@ -0,0 +1,240 @@
+        {messageType:"NoTopLevelSignature", description: "This check will look for function declarations without a signature. We want our readers to understand our code. Adding a signature is a part of this. This check will skip definitions in let statements."},
+        {messageType:"NoUncurriedPrefix", description: "It is unneeded to use an operator in prefix notation when you apply both arguments directly. This check will look for these kind of usages"},
+        {messageType:"RedefineVariable", description: "You should not redefine a variable in a new lexical scope. This is confusing and may lead to bugs."},
+        {messageType:"UnnecessaryListConcat", description: "Yu should not use List.concat to concatenate literal lists. Just join the lists together."},
```
`Yu` typo.

stil4m commented on src/elmCodeAction.ts:
```diff
@@ -125,16 +171,126 @@ function replaceSuggestedVariable(msg : string[]) {
+		vscode.window.showInformationMessage('Language is not Elm');
+    return;
+  }
+	let fullParens = editor.document.getText(range);
```
Indent seems a bit off

hakonrossebo replied:
Seemed ok in my code, but reindented. Fixed in next update

stil4m commented:
In addition thanks for the hard work. I have 3 additional notes:

* What happens if you start `elm-analyse` from different VSCode instances?
* For an experimental feature, wouldnt it be more better to ship the bare minimum, meaning that users have to start `elm-analyse` by hand for now? For the real deal, of course starting `elm-analyse` in the editor would be nicer.
* I see descriptions for all messages, in the near future I want to serve these via a REST call from elm-analyse. Then there will be consistency in all the messages in all the editors, which I think is nice (consistency for the win). Thus keep in mind that I might request to change this structure.

hakonrossebo commented:
Yes, I think it's better to separate the process handling logic for handling the elm-analyse process. I'll see what I can do there. Also, I agree that it may be better to have to have the elm-analyse process be started externally by default for now.

As for starting the elm-analyse from another VSCode instance, it would result in an error saying a process is already running on that port.

For the message descriptions, i only copied it for now to have some descriptions ready for the users - I agree - it's much better if elm-analyse can provide this information.

I'll look into the process handling and get back to you when things are ready.

hakonrossebo commented:
I've refactored code and done some improvements to the process handling. The elm-analyse process seems to start fine in Windows and Linux now. We need to improve Linux process handling, but I think this should be handled in a separate issue.

I've also removed the "experimental" text from readme/pr title as I think this should work ok now. Ready to have the PR merged, but will still wait until @stil4m is ready.

Krzysztof-Cieslak commented:
Can you please rebase it, resolve conflicts, and clean history a bit?

Krzysztof-Cieslak commented:
OK, I'm back, alive. @hakonrossebo, what's the status for this? Ready to merge?

hakonrossebo commented:
Hi, I was not able to compile my correctly on master, but these issues has been discussed and resolved by @littleStudent on Slack and #150. Awaiting PR with typescript-corrections from @littleStudent

littleStudent commented:
here it is. Untouched since the last time we talked https://github.com/Krzysztof-Cieslak/vscode-elm/pull/152

Krzysztof-Cieslak commented:
merged #152

hakonrossebo commented:
Now, I've rebased, cleaned up history and verified it's working ok. Ready to merge.

Krzysztof-Cieslak commented:
Great! :)
